ruby/ext/google/protobuf_c/storage.c

```diff
@@ -484,9 +565,26 @@ void layout_set(MessageLayout* layout,
                 void* storage,
                 const upb_fielddef* field,
                 VALUE val) {
-  void* memory = ((uint8_t *)storage) +
-      layout->offsets[upb_fielddef_index(field)];
-  if (is_map_field(field)) {
+  void* memory = slot_memory(layout, storage, field);
+  uint32_t* oneof_case = slot_oneof_case(layout, storage, field);
+
+  if (upb_fielddef_containingoneof(field)) {
+    if (val == Qnil) {
+      // Assigning nil to a oneof field clears the oneof completely.
+      *oneof_case = 0;
+      memset(memory, 0, NATIVE_SLOT_MAX_SIZE);
+    } else {
+      // Set the oneof case *first* in case a GC is triggered during
```

Comment:
Excellent catch. I've simplified this based on the guarantee provided by `native_slot_set()` (not explicitly documented before) that all calls into the Ruby VM occur before the memory slot is altered. Based on this, we can set the oneof case immediately after `native_slot_set()` returns and be guaranteed that from the point of view of the Ruby VM, both the memory slot and the oneof case change atomically. Added a comment to explain.